README.md
# Introduction

[![Streamlit App](https://static.streamlit.io/badges/streamlit_badge_black_white.svg)](https://share.streamlit.io/pareyesv/streamlit-urbana)


Streamlit demo for spatial data

# Getting Started

To run the app, do

```console
make run-app
```

and follow the instructions.

# Contribute

## Init

To start, please create a virtual environment. The `Makefile` assumes the name of the virtualenv is `streamlit-urbana`. Then, activate the environment.

Then, run:

```console
make init
```

Then, install the rest of the dependencies as suggested in [Dependencies](#dependencies).

## Dependencies

To add/change a dependency, please edit the file `requirements.in`.

Then, run

```console
pip-compile
```

The command abobe will update the `requirements.txt` accordingly.

If the dependency is related to development, please edit the file `requirements-dev.in`.
Then, run

```console
pip-compile requirements-dev.in
```

The command will update the file `requirements-dev.txt`.

Then, to update the environment with the new changes in `requirements.txt` (or `requirements-dev.txt`), run

```console
pip-sync requirements.txt requirements-dev.txt
```

All the steps above can be done by using the `Makefile`:

```console
make update-env
```

## Run the app

To run the app, do:

```console
make run-app
```

## Commit conventions

This project adheres to the [Conventional commits](https://www.conventionalcommits.org/) conventions. Consider using

```console
cz commit
```

to get help writting the message of the commit. For more info, see [commitizen](https://commitizen-tools.github.io/commitizen/).

## Bump version

If you follow the [commit conventions](#commit-conventions), [commitizen](https://commitizen-tools.github.io/commitizen/) will automatically bump the version according to semantic versioning.


CHANGELOG.md
## v0.3.0 (2022-08-08)

### Feat

- set default elevation and color features

## v0.2.1 (2021-07-26)

### Refactor

- change var names and structure

## v0.2.0 (2021-07-05)

### Feat

- manage data for pydeck map


app.py
"""Demo app

Ref:
- https://docs.streamlit.io/en/stable/getting_started.html
- https://deckgl.readthedocs.io/en/latest/gallery/geojson_layer.html
"""

import streamlit as st
import pandas as pd
import pydeck as pdk
from __version__ import version

st.title("InsideAirBnB data for Barcelona")


# data
@st.cache(allow_output_mutation=True)
def get_data():
    """Load and merge data.

    Returns:
        dict: {"describe": pandas decribe() of the features,
               "geojson": geojson formatted dictionary}
    """

    import geopandas as gpd
    import json

    YEAR = 2017
    MONTH = 5
    URL_GEO = "https://raw.githubusercontent.com/egregorimar/urbana/master/data/interim/sections_geo.json"
    URL_DATA = "https://raw.githubusercontent.com/egregorimar/urbana/master/data/interim/sections_{}_{:02d}.csv".format(
        YEAR, MONTH
    )

    df_geo = gpd.read_file(URL_GEO)
    df_geo.set_index("Tag", inplace=True)

    df = pd.read_csv(URL_DATA)
    df.set_index("Tag", inplace=True)

    full_df = df_geo.merge(df, on="Tag")

    return {"describe": df.describe(), "geojson": json.loads(full_df.to_json())}


bcn_data = get_data()
bcngeojson = bcn_data["geojson"]
features = bcn_data["describe"].columns.values

# select features for elevation & color
option_elevation = st.selectbox(
    "Which feature do you like to visualize as elevation?", features
)
option_color = st.selectbox(
    "Which feature do you like to visualize as color?", features
)

# plot
BCN_INITIAL_VIEW_STATE = pdk.ViewState(
    latitude=41.38879, longitude=2.15899, zoom=11, max_zoom=16, pitch=45, bearing=0
)

bcngeojson = pdk.Layer(
    "GeoJsonLayer",
    data=bcngeojson,
    opacity=0.8,
    stroked=False,
    filled=True,
    extruded=True,
    wireframe=True,
    get_elevation=f"properties.{option_elevation}",
    elevation_scale=bcn_data["describe"].loc["max"].max()
    / bcn_data["describe"][option_elevation]["max"],
    get_fill_color=f"""[255, (1 - properties.{option_color} / {bcn_data["describe"][option_color]["max"]}) * 255, 255]""",
    get_line_color=[255, 255, 255],
)

r = pdk.Deck(layers=[bcngeojson], initial_view_state=BCN_INITIAL_VIEW_STATE)
st.pydeck_chart(r)

# credits and version
st.markdown(
    f"Based on [streamlit-urbana](https://github.com/pareyesv/streamlit-urbana), v{version}"
)


streamlit_app.py
"""Demo app

Ref:
- https://docs.streamlit.io/en/stable/getting_started.html
- https://deckgl.readthedocs.io/en/latest/gallery/geojson_layer.html
"""

import json

import geopandas as gpd
import pandas as pd
import pydeck as pdk
import streamlit as st

from __version__ import version

YEAR = 2017
MONTH = 5
URL_GEO = "https://raw.githubusercontent.com/egregorimar/urbana/master/data/interim/sections_geo.json"
URL_DATA = "https://raw.githubusercontent.com/egregorimar/urbana/master/data/interim/sections_{}_{:02d}.csv".format(
    YEAR, MONTH
)

INITIAL_LAT = 41.38879
INITIAL_LON = 2.15899
CITY_NAME = "Barcelona"

DEFAULT_OPTION_ELEVATION = "Airbnb_Number"
DEFAULT_OPTION_COLOR = "POI_Restaurants_Hotels"

st.title(f"InsideAirBnB data in {CITY_NAME}")


# data
@st.cache(allow_output_mutation=True)
def get_data(url_geo: str, url_data: str) -> dict:
    """Load and merge data.
    Returns:
        dict: {"describe": pandas decribe() of the features,
               "geojson": geojson formatted dictionary}
    """

    df_geo = gpd.read_file(url_geo)
    df_geo.set_index("Tag", inplace=True)

    df = pd.read_csv(url_data)
    df.set_index("Tag", inplace=True)

    full_df = df_geo.merge(df, on="Tag")

    return {"describe": df.describe(), "geojson": json.loads(full_df.to_json())}


city_data = get_data(url_geo=URL_GEO, url_data=URL_DATA)
features = city_data["describe"].columns.values

# select features for elevation & color
option_elevation = st.selectbox(
    "Which feature do you like to visualize as elevation?",
    features,
    index=features.tolist().index(DEFAULT_OPTION_ELEVATION),
)
option_color = st.selectbox(
    "Which feature do you like to visualize as color?",
    features,
    index=features.tolist().index(DEFAULT_OPTION_COLOR),
)

# plot
INITIAL_VIEW_STATE = pdk.ViewState(
    latitude=INITIAL_LAT,
    longitude=INITIAL_LON,
    zoom=11,
    max_zoom=16,
    pitch=45,
    bearing=0,
)

geojson_layer = pdk.Layer(
    "GeoJsonLayer",
    data=city_data["geojson"],
    # opacity=0.8,
    stroked=False,
    filled=True,
    extruded=True,
    wireframe=True,
    get_elevation=f"properties.{option_elevation}",
    elevation_scale=city_data["describe"].loc["max"].max()
    / city_data["describe"][option_elevation]["max"],
    get_fill_color=f"""[255, (1 - properties.{option_color} / {city_data["describe"][option_color]["max"]}) * 255, 255]""",
    get_line_color=[255, 255, 255],
)

st.pydeck_chart(pdk.Deck(layers=[geojson_layer], initial_view_state=INITIAL_VIEW_STATE))

# credits and version
st.markdown(
    f"Based on [streamlit-urbana](https://github.com/pareyesv/streamlit-urbana), v{version}"
)
